// @peaq-network/types/src/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.options = exports.defaultOptions = exports.typesBundleForPolkadotApps = exports.mRpc = exports.mTypes = void 0;
const tslib_1 = require("tslib");
const definitions = tslib_1.__importStar(require("./interfaces/definitions"));
require("./interfaces/augment-api");
require("./interfaces/augment-api-consts");
require("./interfaces/augment-api-errors");
require("./interfaces/augment-api-events");
require("./interfaces/augment-api-query");
require("./interfaces/augment-api-rpc");
require("./interfaces/augment-api-tx");
require("./interfaces/augment-types");
require("./interfaces/augment-api-runtime");
exports.mTypes = Object.values(definitions).reduce((res, { types }) => (Object.assign(Object.assign({}, res), types)), {});
exports.mRpc = {
    PeaqDID: {
        read_attribute: {
            description: "Read attribute",
            params: [
                {
                    name: "did_account",
                    type: "AccountId",
                },
                {
                    name: "name",
                    type: "Bytes",
                },
                {
                    name: "at",
                    type: "Option<BlockHash>",
                },
            ],
            type: "Attribute<BlockNumber, Moment>",
        },
    },
};
exports.typesBundleForPolkadotApps = {
    types: [
        {
            minmax: [0, undefined],
            types: exports.mTypes
        }
    ],
    rpc: exports.mRpc
};
exports.defaultOptions = {
    types: exports.mTypes,
    rpc: exports.mRpc
};
const options = (_a = {}) => {
    var { types = {}, rpc = {} } = _a, otherOptions = tslib_1.__rest(_a, ["types", "rpc"]);
    return (Object.assign({ types: Object.assign(Object.assign({}, exports.mTypes), types), rpc: Object.assign(Object.assign({}, exports.mRpc), rpc) }, otherOptions));
};
exports.options = options;
