// @peaq-network/sdk/src/utils/errors.js
"use strict";
// DID custom errors
Object.defineProperty(exports, "__esModule", { value: true });
exports.StorageSeedError = exports.StorageAddressError = exports.ItemError = exports.ItemTypeError = exports.StorageError = exports.RemoveDidError = exports.DidNotFoundError = exports.NoCustomFieldsError = exports.UpdateDidError = exports.ReadDidError = exports.AddressError = exports.SeedError = exports.NameError = exports.CreateDidError = exports.GenerateDidError = void 0;
class GenerateDidError extends Error {
    constructor(message) {
        super(message);
        this.name = 'GenerateDidError';
        Object.setPrototypeOf(this, new.target.prototype);
    }
}
exports.GenerateDidError = GenerateDidError;
class CreateDidError extends Error {
    constructor(message) {
        super(message);
        this.name = 'CreateDidError';
        Object.setPrototypeOf(this, new.target.prototype);
    }
}
exports.CreateDidError = CreateDidError;
class NameError extends CreateDidError {
    constructor(message) {
        super(message);
        this.name = 'NameError';
        Object.setPrototypeOf(this, new.target.prototype);
    }
}
exports.NameError = NameError;
class SeedError extends CreateDidError {
    constructor(message) {
        super(message);
        this.name = 'SeedError';
        Object.setPrototypeOf(this, new.target.prototype);
    }
}
exports.SeedError = SeedError;
class AddressError extends CreateDidError {
    constructor(message) {
        super(message);
        this.name = 'AddressError';
        Object.setPrototypeOf(this, new.target.prototype);
    }
}
exports.AddressError = AddressError;
class ReadDidError extends Error {
    constructor(message) {
        super(message);
        this.name = 'ReadDidError';
        Object.setPrototypeOf(this, new.target.prototype);
    }
}
exports.ReadDidError = ReadDidError;
class UpdateDidError extends Error {
    constructor(message) {
        super(message);
        this.name = 'UpdateDidError';
        Object.setPrototypeOf(this, new.target.prototype);
    }
}
exports.UpdateDidError = UpdateDidError;
class NoCustomFieldsError extends UpdateDidError {
    constructor(message) {
        super(message);
        this.name = 'NoCustomFieldsError';
        Object.setPrototypeOf(this, new.target.prototype);
    }
}
exports.NoCustomFieldsError = NoCustomFieldsError;
class DidNotFoundError extends UpdateDidError {
    constructor(message) {
        super(message); // Pass the message to the parent Error class
        this.name = 'DidNotFoundError'; // Set the name property to the custom error type
        Object.setPrototypeOf(this, new.target.prototype); // Restore the prototype chain
    }
}
exports.DidNotFoundError = DidNotFoundError;
class RemoveDidError extends Error {
    constructor(message) {
        super(message); // Pass the message to the parent Error class
        this.name = 'RemoveDidError'; // Set the name property to the custom error type
        Object.setPrototypeOf(this, new.target.prototype); // Restore the prototype chain
    }
}
exports.RemoveDidError = RemoveDidError;
// Storage Custom Errors
class StorageError extends Error {
    constructor(message) {
        super(message); // Pass the message to the parent Error class
        this.name = 'StorageError'; // Set the name property to the custom error type
        Object.setPrototypeOf(this, new.target.prototype); // Restore the prototype chain
    }
}
exports.StorageError = StorageError;
// Define CreateDidError as a child of the built-in Error class
class ItemTypeError extends StorageError {
    constructor(message) {
        super(message); // Pass the message to the parent Error class
        this.name = 'ItemTypeError'; // Set the name property to the custom error type
        Object.setPrototypeOf(this, new.target.prototype); // Restore the prototype chain
    }
}
exports.ItemTypeError = ItemTypeError;
// Define CreateDidError as a child of the built-in Error class
class ItemError extends StorageError {
    constructor(message) {
        super(message); // Pass the message to the parent Error class
        this.name = 'ItemError'; // Set the name property to the custom error type
        Object.setPrototypeOf(this, new.target.prototype); // Restore the prototype chain
    }
}
exports.ItemError = ItemError;
// Define CreateDidError as a child of the built-in Error class
class StorageAddressError extends StorageError {
    constructor(message) {
        super(message); // Pass the message to the parent Error class
        this.name = 'StorageAddressError'; // Set the name property to the custom error type
        Object.setPrototypeOf(this, new.target.prototype); // Restore the prototype chain
    }
}
exports.StorageAddressError = StorageAddressError;
// Define CreateDidError as a child of the built-in Error class
class StorageSeedError extends StorageError {
    constructor(message) {
        super(message); // Pass the message to the parent Error class
        this.name = 'StorageSeedError'; // Set the name property to the custom error type
        Object.setPrototypeOf(this, new.target.prototype); // Restore the prototype chain
    }
}
exports.StorageSeedError = StorageSeedError;
